File: 611_Valid_Triangle_Number.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution{
public:
    int triangleNumber(vector<int>& nums){
        /*binary search O(n^2*log(n)) : 34% */
        int n = nums.size(); int cnt = 0;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 2; ++i){
            for (int j = i + 1; j < n - 1; ++j){
                int right = n, left = j + 1;
                int target = nums[i] + nums[j];
                while (left < right){
                    int mid = (left + right) / 2;
                    if (nums[mid] < target) left = mid + 1;
                    else right = mid;
                }
                cnt += right - j - 1;
            }
        }
        return cnt;
    }
};

int main(){
    Solution* solution = new Solution();
    vector<int> nums{2, 2, 3, 4};
    cout << solution->triangleNumber(nums) << endl;
    return 0;
}

File: 611_Valid_Triangle_Number.py
class Solution(object):
    def triangleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Time: O(N^2log(N))
        # Space: O(1)
        if len(nums) < 3: return 0

        nums.sort()

        cnt = 0
        for i in range(len(nums) - 2):
            for j in range(i + 1, len(nums) - 1):
                l, k = j + 1, len(nums)
                while l < k:
                    m = l + (k - l) // 2
                    if nums[m] < nums[i] + nums[j]:
                        l = m + 1
                    else:
                        k = m
                cnt += k - j - 1
        return cnt

    def triangleNumber2(self, nums):
        # Time: O(N^2)
        # Space: O(1)
        nums.sort()
        cnt = 0
        for k in range(2, len(nums))[::-1]:
            i, j = 0, k - 1
            while i < j:
                if nums[i] + nums[j] > nums[k]:
                    cnt += j - i
                    j -= 1
                else:
                    i += 1
        return cnt


solution = Solution()
print(solution.triangleNumber2([2,2,3,4]))